Order capturing changes

philbirt opened the pull request:
* Config for auto_capture_exchanges
** I felt like this should be independent of auto_capturing orders. For instance, we will have auto_capture false in Bonobos, but want to auto capture for exchanges.

* Partial capture updates payment amount
** This is a change we are getting from spree 2-4 anyway, but needed it for current work until we upgrade to Solidus

* Clean up shipment state machine
** We needed to hook into the conditions of how a shipment can go to the ready state, but this was impossible. Also this gets us a little less dependent on `Shipment#determine_state` which seems like one of the most insane methods I've seen.

cc @jordan-brough @athal7 @cbrunsdon @gmacdougall @jhawthorn

jordan-brough commented on core/app/models/spree/payment/processing.rb:
```diff
@@ -52,6 +52,7 @@ def capture!(amount=nil)
 
           money = ::Money.new(amount, Spree::Config[:currency])
           capture_events.create!(amount: money.to_f)
+          update_attributes(amount: captured_amount)
```
can we use the bang version `update_attributes!` here?

athal7 commented on core/app/models/spree/shipment.rb:
```diff
@@ -62,12 +56,7 @@ class Shipment < ActiveRecord::Base
-          shipment.determine_state(shipment.order) == :ready
-        }
-        transition from: :canceled, to: :pending, if: lambda { |shipment|
-          shipment.determine_state(shipment.order) == :ready
```
haha, this never worked :(

jordan-brough replied:
ha, funny

philbirt replied:
yep.

jordan-brough commented:
Looking good to me.  We're planning on monkey-patching the `can_transition_from_pending_to_ready?` method, right?  Should we instead add some kind of preference like `require_payment_to_ship` or something?

athal7 commented on core/app/models/spree/shipment.rb:
```diff
@@ -81,6 +70,18 @@ class Shipment < ActiveRecord::Base
+    end
+
+    def can_transition_from_pending_to_ready?
+      order.can_ship? && !inventory_units.any?(&:backordered?) && order.paid?
```
if you break out order.paid? into something separate method you'll have an easier time overriding that piece for capturing at dispatch.

athal7 commented:
looks great!

philbirt commented:
* Added a configuration to require/not require payment for a shipment to be ready.

* Added OrderCapturing to allow for processing full or partial multiple payments for an order
** Hook for sorting which payments you want to run first in your own store

* Added a rake task that invokes OrderCapturing#capture_payments given a time window

philbirt commented on core/app/models/spree/payment.rb:
```diff
@@ -184,7 +188,7 @@ def create_payment_profile
 
       def invalidate_old_payments
-        order.payments.with_state('checkout').where("id != ?", self.id).each do |payment|
+        order.payments.with_state('checkout').where(payment_method: payment_method).where("id != ?", self.id).each do |payment|
```
Note: This is going to go in with the store credits work. This allows us to have multiple payments, and was needed to test the OrderCapturing class in a reasonable manner.

jordan-brough commented on core/app/models/spree/order_capturing.rb:
```diff
@@ -0,0 +1,32 @@
+        payment.capture!((amount * 100).to_i)
+        uncaptured_amount -= amount
+      end
+    end
```
Should we call `@order.update!` before exiting the OrderMutex?

philbirt replied:
Added

jordan-brough commented on core/lib/spree/core/unreturned_item_charger.rb:
```diff
@@ -34,6 +34,7 @@ def charge_for_items
 
       new_order.contents.approve(name: self.class.name)
       new_order.reload.complete!
+      new_order.payments.map(&:capture!) if (Spree::Config[:auto_capture_exchanges] && !Spree::Config[:auto_capture])
```
Would it make sense to use `Spree::OrderCapturing` here instead of `map(&:capture!)`?

philbirt replied:
Sounds good to me

jordan-brough commented:
:+1:

jordan-brough commented on core/app/models/spree/order_capturing.rb:
```diff
@@ -0,0 +1,34 @@
+class Spree::OrderCapturing
+  class_attribute :eligible_payments
```
would this be better names as `sorted_payment_method_classes` or something like that? (less verbose if possible...)  Since it's not really about "eligibleness"

cbrunsdon commented:
Okay, I've worked through a lot of stuff with this over slack with @philbirt so I'm :+1:  for it now.

Things to work on in the future though:
 * Split up functionality into multiple PR's ( I don't think the order capture stuff had to go next to the cleanups of the shipment state machine )
 * Strive to stub as little as we can in our tests
 * Examine OrderCapture in $X months and figure out if there was something we could have done better to make it clearer or more extensible

jordan-brough commented:
:+1:
